### PR TITLE
feature: add providers api key args

### PR DESCRIPTION
**Pull request by 1l0c**
Enhance Subfinder Agent to Support All Data Provider API Keys 

Changes:

- Introduce ProviderConfigManager class for centralized management of provider API keys
- Expand provider support from VirusTotal-only to all supported providers
- Add a comprehensive test suite covering edge cases
- Refactor and optimize provider API key configuration setup logic
- Bump subfinder to v2.12.0

### CODE REVIEW

File: agent/provider_config_manager.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+    def __init__(self, config_path: str = CONFIG_PATH):
+        self.config_path = config_path
+
+    def add_provider_key(self, provider_name: str, api_key: str) -> None:
+        """
+        Add an API key for a specific provider.
+        """
+
+        if not provider_name:
```
Comment (yasser-abassi) [inline agent/provider_config_manager.py]: please all if checks should be explicit

File: agent/config.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+from rich import logging as rich_logging
+
+
+logging.basicConfig(
+    format="%(message)s",
+    datefmt="[%X]",
+    handlers=[rich_logging.RichHandler(rich_tracebacks=True)],
+    level="INFO",
+    force=True,
+)
```
Comment (3asm) [inline agent/config.py]: Not needed, will inherit from existing config.

File: agent/config.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
+import logging
```
Comment (3asm) [inline agent/config.py]: Missing docstring.

File: agent/provider_config_manager.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
+from agent.config import logger, CONFIG_PATH
```
Comment (3asm) [inline agent/provider_config_manager.py]: Per style guide, don't import classes directly.

Comment (3asm) [reply agent/provider_config_manager.py]: create a logger here, don't import it.

Comment (3asm) [reply agent/provider_config_manager.py]: Why do we need CONFIG_PATH from a different module. The complexity of the code don't require it.

File: agent/provider_config_manager.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import pathlib
+
+
+class ProviderConfigManager:
+    """
+    Manages API keys for providers in the Subfinder configuration file.
+    """
+
+    def __init__(self, config_path: str = CONFIG_PATH):
+        self.config_path = config_path
```
Comment (3asm) [inline agent/provider_config_manager.py]: Private if not part of the contract.

File: agent/provider_config_manager.py (hunk 1 above)
Comment (3asm) [inline agent/provider_config_manager.py]: Use explicit check per style guide.

File: agent/provider_config_manager.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        """
+
+        if not provider_name:
+            logger.error("Provider name cannot be empty.")
+            return
+        if not api_key:
+            logger.error("API key cannot be empty for provider '%s'.", provider_name)
+            return
+
+        self.__save_provider_key(provider_name, api_key)
```
Comment (3asm) [inline agent/provider_config_manager.py]: Single `_`. No need for 2.

File: agent/provider_config_manager.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+    def __save_provider_key(self, provider_name: str, api_key: str) -> None:
+        """
+        Save an API key for a given provider in the Subfinder configuration file.
+
+        Args:
+            provider_name (str): The provider name (e.g., 'virustotal', 'fofa').
+            api_key (str): The API key to add.
+        """
+        yaml = ruamel.yaml.YAML(typ="safe")
```
Comment (3asm) [inline agent/provider_config_manager.py:32]: Did you ruamel in the deps?

File: agent/subfinder_agent.py [hunk 8]
```diff
@@ -14,45 +9,11 @@
 
 from agent import subfinder
 
+from agent.config import logger, STORAGE_NAME, PROVIDER_ARG_MAP
+from agent.provider_config_manager import ProviderConfigManager
```
Comment (3asm) [inline agent/subfinder_agent.py]: Don't import classes directly.

File: agent/subfinder_agent.py [hunk 9]
```diff
@@ -14,45 +9,11 @@
...
 
-logging.basicConfig(
```
Comment (3asm) [inline agent/subfinder_agent.py:18]: Please don't remove this.

File: agent/subfinder_agent.py [hunk 10]
```diff
@@ -65,14 +26,49 @@ def __init__(
...
+            provider_api_key = args.get(arg)
+            if not provider_api_key:
+                logger.debug(
+                    "No API key provided for provider argument '%s'; skipping.",
+                    arg,
+                )
+                continue
+
+            provider_name = PROVIDER_ARG_MAP.get(arg)
+            if not provider_name:
```
Comment (3asm) [inline agent/subfinder_agent.py]: Please use explicit check.

File: tests/provider_config_manager_test.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
+"""Unit tests for ProviderConfigManager."""
+
+import pathlib
+
+import pytest
+import ruamel.yaml
+from pyfakefs import fake_filesystem_unittest
+from pytest_mock import plugin
+
+from agent.provider_config_manager import ProviderConfigManager
```
Comment (3asm) [inline tests/provider_config_manager_test.py]: Don't import classes directly.

**Review by 3asm — CHANGES_REQUESTED** [review]

File: agent/provider_config_manager.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+    def add_provider_key(self, provider_name: str, api_key: str) -> None:
+        """
+        Add an API key for a specific provider.
+        """
+
+        if not provider_name:
+            logger.error("Provider name cannot be empty.")
+            return
+        if not api_key:
```
Comment (yami-ost) [inline agent/provider_config_manager.py]: use explicit checks

File: agent/provider_config_manager.py (hunk 1 above)
Comment (yasser-abassi) [reply agent/provider_config_manager.py]: here is the style guide: https://internal-docs.ostorlab.app/docs/dev/clean_code/

File: agent/subfinder_agent.py [hunk 13]
```diff
@@ -65,14 +26,49 @@ def __init__(
...
+        # Update providers configuration with API keys
+        self.update_providers_api_keys(self.args)
+
         self._use_all_sources: bool = self.args.get("use_all_sources") or False
         self._active_only: bool = self.args.get("active_only") or False
         agent_persist_mixin.AgentPersistMixin.__init__(self, agent_settings)
 
+    def update_providers_api_keys(self, args: dict[str, str]) -> None:
+
+        # Update providers configuration with API keys from the provided arguments.
```
Comment (yami-ost) [inline agent/subfinder_agent.py]: we use docStrings not comments

File: agent/provider_config_manager.py (hunk 1 above)
Comment (1l0c) [reply agent/provider_config_manager.py]: Thanks Yasser for sharing the style guide

File: agent/provider_config_manager.py (hunk 4 above)
Comment (1l0c) [reply agent/provider_config_manager.py]: I moved CONFIG_PATH and other configuration constants into a separate config module to centralize configuration and avoid hard-coding paths across modules. This makes future changes or overrides easier without touching the business logic.

File: agent/provider_config_manager.py (hunk 7 above)
Comment (1l0c) [reply agent/provider_config_manager.py:32]: Yes, it was already included in the dependencies.

File: agent/config.py (hunk 3 above)
Comment (burogurama) [inline agent/config.py]: Please add a module docstring
```suggestion
"""..."""

import logging
```

File: agent/config.py (hunk 2 above)
Comment (burogurama) [inline agent/config.py]:
```suggestion
```
Logging is already configured in the main module. Please remove.

File: agent/provider_config_manager.py (hunk 4 above)
Comment (burogurama) [inline agent/provider_config_manager.py]: Please do module docsring.

**Review by burogurama — CHANGES_REQUESTED** [review]

File: agent/subfinder_agent.py [hunk 14]
```diff
@@ -65,14 +26,49 @@ def __init__(
...
+
+        providers_arg_names = PROVIDER_ARG_MAP.keys()
+        logger.info("Starting update of providers configuration with API keys.")
+
+        for arg in args:
+            if arg not in providers_arg_names:
+                continue
```
Comment (MouadAO) [inline agent/subfinder_agent.py]: I think no need for this variable `providers_arg_names`:

you can do this : 


```suggestion
        logger.info("Starting update of providers configuration with API keys.")

        for arg in args:
            if arg not in PROVIDER_ARG_MAP:
                continue
```

File: agent/subfinder_agent.py [hunk 15]
```diff
@@ -65,14 +41,46 @@ def __init__(
...
+    def update_providers_api_keys(self, args: dict[str, Any]) -> None:
+        """Updates providers configuration with API keys from arguments."""
+
+        logger.info("Starting update of providers configuration with API keys.")
+
+        for arg in args:
+            if arg not in PROVIDER_ARG_MAP:
+                continue
+
+            provider_api_key = args.get(arg)
```
Comment (yami-ost) [inline agent/subfinder_agent.py]: Since you already know the key exists, .get() is unnecessary
```suggestion
        for arg, provider_api_key in args.items():
            if arg not in PROVIDER_ARG_MAP:
                continue
```

File: agent/subfinder_agent.py [hunk 16]
```diff
@@ -65,14 +41,46 @@ def __init__(
...
+            if provider_api_key is None or provider_api_key.strip() == "":
```
Comment (yami-ost) [inline agent/subfinder_agent.py]: provider_api_key is type Any and strip is for strings you should avoid type Any if provider_api_key is a string say so else check for it explicitly

**Review by yami-ost — CHANGES_REQUESTED** [review]

File: agent/provider_config_manager.py [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+import pathlib
+
+import ruamel.yaml
+
+from agent import config
+
+logger = logging.getLogger(__name__)
+
+CONFIG_PATH = config.CONFIG_PATH
+PROVIDER_ARG_MAP = config.PROVIDER_ARG_MAP
```
Comment (burogurama) [inline agent/provider_config_manager.py]: You defined them in config, why re-declare them here, just use `config.CONFIG_PATH`, otherwise what's the point of centralizing config.

```suggestion
```

File: agent/provider_config_manager.py [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+
+
+class ProviderConfigManager:
+    """Manages API keys for providers in the Subfinder YAML configuration file."""
+
+    def __init__(self, config_path: str = CONFIG_PATH):
```
Comment (burogurama) [inline agent/provider_config_manager.py]:
```suggestion
    def __init__(self, config_path: str = config.CONFIG_PATH):
```

**Review by burogurama — CHANGES_REQUESTED** [review]

File: agent/provider_config_manager.py [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+        self._config_path = config_path
+
+    def add_provider_key(self, provider_name: str, api_key: str) -> None:
+        """Adds an API key for a specific provider."""
+        if provider_name is None or provider_name.strip() == "":
```
Comment (burogurama) [inline agent/provider_config_manager.py]: provider_name is `str` not `str | None`, so please remove the `is None` check
```suggestion
        if provider_name.strip() == "":
```

File: agent/provider_config_manager.py [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+            logger.error("Provider name cannot be empty.")
+            return
+        if api_key is None or api_key.strip() == "":
+            logger.error("API key cannot be empty for provider '%s'.", provider_name)
+            return
```
Comment (burogurama) [inline agent/provider_config_manager.py]:
```suggestion
        if api_key.strip() == "":
            logger.error("API key cannot be empty for provider '%s'.", provider_name)
            return
```
Same for `api_key`

File: agent/provider_config_manager.py [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+            if api_key not in config[provider_name]:
+                config[provider_name].append(api_key)
+                logger.info("Added API key for provider '%s'.", provider_name)
+            else:
+                logger.info(
+                    "API key for provider '%s' already exists; skipping.", provider_name
+                )
+        else:
+            config[provider_name] = [api_key]
+            logger.info("Created new provider entry '%s' with API key.", provider_name)
```
Comment (burogurama) [inline agent/provider_config_manager.py]: Invert the check to reduce the `if` block nesting.

```suggestion
        if provider_name not in config:
            config[provider_name] = []
            logger.info("Created new provider entry '%s' with API key.", provider_name)
        if api_key not in config[provider_name]:
            config[provider_name].append(api_key)
            logger.info("Added API key for provider '%s'.", provider_name)
        else:
           logger.info(
               "API key for provider '%s' already exists; skipping.", provider_name
           )
```

**Review by burogurama — CHANGES_REQUESTED** [review]

File: agent/subfinder_agent.py [hunk 22]
```diff
@@ -1,58 +1,34 @@
...
-    config = yaml.load(config_path.read_text()) or {}
+logger = logging.getLogger(__name__)
 
-    if "virustotal" in config:
-        if virustotal_key not in config["virustotal"]:
-            config["virustotal"].append(virustotal_key)
-    else:
-        config["virustotal"] = [virustotal_key]
+PROVIDER_ARG_MAP = config.PROVIDER_ARG_MAP
+STORAGE_NAME = config.STORAGE_NAME
```
Comment (burogurama) [inline agent/subfinder_agent.py]: Remove and use `config.* `directly.

File: agent/subfinder_agent.py (hunk 15 above)
Comment (burogurama) [inline agent/subfinder_agent.py]:
```suggestion
        for arg, value in args.items():
            if arg not in PROVIDER_ARG_MAP:
                continue

            provider_api_key = value
```

File: agent/subfinder_agent.py (hunk 16 above)
Comment (burogurama) [reply agent/subfinder_agent.py]: +1

File: agent/config.py [hunk 23]
```diff
@@ -0,0 +1,57 @@
+"""Configuration module for the Subfinder agent."""
+
+STORAGE_NAME = "agent_subfinder_storage"
```
Comment (burogurama) [inline agent/config.py]: Please add explicit typing to global and public attributes.
```suggestion
STORAGE_NAME: str = "agent_subfinder_storage"
```

File: agent/config.py [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+
+CONFIG_PATH = "/root/.config/subfinder/provider-config.yaml"
```
Comment (burogurama) [inline agent/config.py]:
```suggestion
CONFIG_PATH: str = "/root/.config/subfinder/provider-config.yaml"
```

File: agent/config.py [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+
+PROVIDER_ARG_MAP = {
```
Comment (burogurama) [inline agent/config.py]: Add typing hint please

File: agent/provider_config_manager.py [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+
+        config_path = pathlib.Path(self._config_path)
+
+        if config_path.exists() is False:
+            logger.error("Configuration file not found at %s.", config_path)
+            return
+
+        try:
+            loaded_config = yaml.load(config_path.read_text())
+            config = loaded_config if loaded_config is not None else {}
```
Comment (burogurama) [inline agent/provider_config_manager.py:43]: You already imported config as a module, change this variable name or the import name, I suggest importing config as `agent_config`

File: agent/subfinder_agent.py [hunk 27]
```diff
@@ -65,14 +41,46 @@ def __init__(
...
+                )
+                continue
+
+            provider_name = PROVIDER_ARG_MAP.get(arg)
+            if provider_name is None or provider_name.strip() == "":
+                logger.warning(
+                    "No provider mapping found for argument '%s'; skipping.",
+                    arg,
+                )
+                continue
```
Comment (burogurama) [inline agent/subfinder_agent.py]: You already checked that arg is in PROVIDER_ARG_MAP, and the map should not allow None as values, so remove the check.
```suggestion
            provider_name = PROVIDER_ARG_MAP[arg]
```

**Review by burogurama — CHANGES_REQUESTED** [review]

File: agent/provider_config_manager.py (hunk 26 above)
Comment (burogurama) [reply agent/provider_config_manager.py:43]: \+ type config

**Review by 3asm — APPROVED** [review]
